server/src/main/scala/org/http4s/server/websocket/WebSocketBuilder.scala

```diff
@@ -0,0 +1,39 @@
+package org.http4s.server.websocket
+
+import cats._
+import cats.implicits._
+import fs2._
+import org.http4s.websocket.WebsocketBits.WebSocketFrame
+import org.http4s.websocket.{WebSocketContext, Websocket}
+import org.http4s.{AttributeEntry, Headers, Response, Status}
+
+case class WebSocketBuilder[F[_]](send: Stream[F, WebSocketFrame], receive: Sink[F, WebSocketFrame])(
+    implicit F: Monad[F]) {
+
+  private var onNonWebSocketRequest: F[Response[F]] =
+    Response[F](Status.NotImplemented).withBody("This is a WebSocket route.")
+  private var onHandshakeFailure: F[Response[F]] =
+    Response[F](Status.BadRequest).withBody("WebSocket handshake failed.")
+  private var handshakeResponseHeaders: Headers = Headers.empty
```

Comment:
This can be done without vars, see my other comment or look at `BlazeBuilder`: https://github.com/http4s/http4s/blob/master/blaze-server/src/main/scala/org/http4s/server/blaze/BlazeBuilder.scala

Reply:
@rossabaker @aeons @SystemFw thank you for your feedback, I am learning a lot from you guys. I used BlazeBuilder as reference. Still wrapping my head around the (dark) Magic example, but as its now 02:00 for me, I will sure have to return to this tomorrow morning. I hope the current state of the code is good enough to be approved, though.